src/vcTiffConvert.cpp

```diff
@@ -10,157 +10,592 @@
+  uint16 samplesPerPixel;
+  uint16 photometric;
+  uint16 planarconfig;
+  uint16 **ppColourPalette; // Stored RRR...GGG...BBB..., number of colours == 2 ^ bitsPerSample, also samplesPerPixel MUST = 1
+};
+
+struct vcTiffConvertData;
+
+typedef void(*vcTiffImageReaderDestroyFn)(vcTiffConvertData *pData);
+typedef udError(*vcTiffImageReaderReadPixelFn)(vcTiffConvertData *pData, vcTiffPixelData *pPixelData);
+
+struct TileData
+{
+  uint8_t *pRaster;
+  uint64_t currentPixel;
+  udVector2<uint32> tileDimensions;
+  udVector2<uint32_t> currentTileCoords;
+  uint64_t nTilePixels; // Can be less then the size of a tile
+};
+
+struct ScanLineData
+{
+  uint8_t *pRaster;
+  uint64_t currentPixel;
+  uint32 row;
+};
+
+struct vcTiffConvertData
+{
+  TIFF *pTif;
+  vcTiffImageReaderDestroyFn pDestroy;
+  vcTiffImageReaderReadPixelFn pRead;
+
+  vcTiffImageFormat format;
+  udVector2<uint32> imageDimensions;
 
-  virtual ~TiffPixelReader()
+  union
   {
+    TileData tile;
+    ScanLineData scanLine;
+  };
 
-  }
+  uint32_t pointsWritten;
+  uint32_t everyNth;
+  double pointResolution;
+  udDouble3 origin;
+  udConvertCustomItemFlags convertFlags;
 
-  udError Init(TIFF *pTiff)
+  uint32_t currentDirectory;
+  uint32_t directoryCount;
+  uint64_t pointsProcessed;
+};
+
+static void vcTiff_PackSamples(uint8_t *pBuf, uint8_t *pSampleBuf, uint32_t sampleCount, uint32_t sampleBitWidth, uint32_t samplesPerPixel, uint32_t sample)
+{
+  if (sampleBitWidth < 8)
   {
-    m_pTiff = pTiff;
-    return OnInit();
+    UDASSERT(false, "TODO!");
   }
-
-  virtual void ShutDown()
+  else
   {
-    OnShutDown();
+    for (size_t i = 0; i < sampleCount; ++i)
+    {
+      uint8_t *pDest = pBuf + sample + i * ((size_t)samplesPerPixel + sampleBitWidth / CHAR_BIT);
+      uint8_t *pSrc = pSampleBuf + i * sampleBitWidth / CHAR_BIT;
+      memcpy(pDest, pSrc, sampleBitWidth);
+    }
   }
+}
+
+static uint32_t vcTiff_To8BitSample(double x)
+{
+  return uint32_t(udClamp(x, 0.0, 1.0) * 255.);
+}
+
+static uint32_t vcTiff_To8BitSample(uint64_t value, uint32_t nBits)
+{
+  if (nBits < 8)
+    return uint32_t(255ull * value / ((1ull << nBits) - 1));
+  if (nBits > 8)
+    return uint32_t(value >> (nBits - 8));
+  return (uint32_t)value;
+}
 
-  // Returns:
-  //    udE_Success: next pixel set
-  //    udE_NotFound: no more pixels to be found
-  //    udE_*: An error has occured
-  virtual udError GetNextPixel(udDouble3 *pCoords, uint32_t *pPixel) = 0;
+static uint32_t vcTiff_To8BitSamplei(int64_t value, int32_t nBits)
+{
+  if (value < 0)
+    return 0;
+
+  if (nBits < 8)
+    return uint32_t(255ll * value / ((1ll << nBits) - 1));
+  if (nBits > 8)
+    return uint32_t(value >> (nBits - 8));
+  return (uint32_t)value;
+}
 
-  void SetConvertDefaults(udDouble3 const & origin, double pointResolution, enum udConvertCustomItemFlags flags)
+static udError vcTiff_SampleDataToColour(vcTiffImageFormat const & format, vcTiffSampleData const & data, uint32_t *pColour)
```

Comment:
Both `*` and `&` go with the variable name, this happens in a few other functions. You may need to adjust your Visual Studio settings if this is automatically happening.